test/unittests/execute_floating_point_test.cpp

```diff
@@ -66,4 +69,203 @@ TEST(execute_floating_point, f64_add)
+    result_type = static_cast<uint8_t>(TypeParam::dst_valtype);
+    opcode = static_cast<uint8_t>(TypeParam::opcode);
+
+    auto instance = instantiate(parse(wasm));
+
+    // Zero.
+    EXPECT_THAT(execute(*instance, 0, {FloatT{0}}), Result(IntT{0}));
+    EXPECT_THAT(execute(*instance, 0, {-FloatT{0}}), Result(IntT{0}));
+
+    // Something around 0.0.
+    EXPECT_THAT(execute(*instance, 0, {FloatLimits::denorm_min()}), Result(IntT{0}));
+    EXPECT_THAT(execute(*instance, 0, {-FloatLimits::denorm_min()}), Result(IntT{0}));
+
+    // Something smaller than 2.0.
+    EXPECT_THAT(execute(*instance, 0, {std::nextafter(FloatT{2}, FloatT{0})}), Result(IntT{1}));
+
+    // Something bigger than -1.0.
+    EXPECT_THAT(execute(*instance, 0, {std::nextafter(FloatT{-1}, FloatT{0})}), Result(IntT{0}));
+
+    {
+        // BOUNDARIES OF DEFINITION
+        //
+        // Here we want to identify and test the boundary values of the defined behavior of the
+        // trunc instructions. For undefined results the execution must trap.
+        // Note that floating point type can represent any power of 2.
+
+        using expected_boundaries = trunc_boundaries<FloatT, IntT>;
+
+        // For iN with max value 2^N-1 the float(2^N) exists and trunc(float(2^N)) to iN
+        // is undefined.
+        const auto upper_boundary = std::pow(FloatT{2}, FloatT{IntLimits::digits});
+        EXPECT_EQ(upper_boundary, expected_boundaries::upper);
+        EXPECT_THAT(execute(*instance, 0, {upper_boundary}), Traps());
+
+        // But the trunc() of the next float value smaller than 2^N is defined.
+        // Depending on the resolution of the floating point type, the result integer value may
+        // be other than 2^(N-1).
+        const auto max_defined = std::nextafter(upper_boundary, FloatT{0});
+        const auto max_defined_int = static_cast<IntT>(max_defined);
+        EXPECT_THAT(execute(*instance, 0, {max_defined}), Result(max_defined_int));
+
+        // The lower boundary is:
+        // - for signed integers: -2^N - 1,
+        // - for unsigned integers: -1.
+        // However, the -2^N - 1 may be not representative in a float type so we compute it as
+        // floor(-2^N - epsilon).
+        const auto min_defined_int = IntLimits::min();
+        const auto lower_boundary =
+            std::floor(std::nextafter(FloatT{min_defined_int}, -FloatLimits::infinity()));
+        EXPECT_EQ(lower_boundary, expected_boundaries::lower);
+        EXPECT_THAT(execute(*instance, 0, {lower_boundary}), Traps());
+
+        const auto min_defined = std::nextafter(lower_boundary, FloatT{0});
+        EXPECT_THAT(execute(*instance, 0, {min_defined}), Result(min_defined_int));
+    }
+
+    {
+        // NaNs.
+        EXPECT_THAT(execute(*instance, 0, {FloatLimits::quiet_NaN()}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {FloatLimits::signaling_NaN()}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {nan<FloatT>()}), Traps());
```

Comment:
Can there be a NaN with sign bit set (`-nan<FloatT<()`)?

Reply:
Yes, it can be.

Reply:
Added test cases for these.